infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServer.java

```diff
@@ -173,226 +182,265 @@ public void handleTailRequest(CorfuPayloadMsg<TailsRequest> msg, ChannelHandlerC
+        }
+
+        HeaderMsg responseHeader;
+
         if (ex.getCause() instanceof WrongEpochException) {
             WrongEpochException wee = (WrongEpochException) ex.getCause();
-            r.sendResponse(ctx, msg, new CorfuPayloadMsg<>(CorfuMsgType.WRONG_EPOCH, wee.getCorrectEpoch()));
+            responseHeader = getHeaderMsg(req.getHeader(), false, true);
+            r.sendResponse(getResponseMsg(responseHeader, getWrongEpochErrorMsg(wee.getCorrectEpoch())), ctx);
         } else if (ex.getCause() instanceof OverwriteException) {
             OverwriteException owe = (OverwriteException) ex.getCause();
-            r.sendResponse(ctx, msg, CorfuMsgType.ERROR_OVERWRITE
-                    .payloadMsg(owe.getOverWriteCause().getId()));
+            responseHeader = getHeaderMsg(req.getHeader(), false, true);
+            r.sendResponse(getResponseMsg(responseHeader, getOverwriteErrorMsg(owe.getOverWriteCause().getId())), ctx);
         } else if (ex.getCause() instanceof TrimmedException) {
-            r.sendResponse(ctx, msg, CorfuMsgType.ERROR_TRIMMED.msg());
+            responseHeader = getHeaderMsg(req.getHeader(), false, false);
+            r.sendResponse(getResponseMsg(responseHeader, getTrimmedErrorMsg()), ctx);
         } else {
-            r.sendResponse(ctx, msg, CorfuMsgType.ERROR_SERVER_EXCEPTION.payloadMsg(new ExceptionMsg(ex)));
+            responseHeader = getHeaderMsg(req.getHeader(), false, true);
+            r.sendResponse(getResponseMsg(responseHeader, getUnknownErrorMsg(ex)), ctx);
             throw new LogUnitException(ex);
         }
     }
 
     /**
      * Service an incoming write request.
      */
-    @ServerHandler(type = CorfuMsgType.WRITE)
-    public void write(CorfuPayloadMsg<WriteRequest> msg, ChannelHandlerContext ctx, IServerRouter r) {
-        LogData logData = (LogData) msg.getPayload().getData();
-        log.debug("log write: type: {}, address: {}, streams: {}", logData.getType(),
-                logData.getToken(), logData.getBackpointerMap());
+    @RequestHandler(type = PayloadCase.WRITE_LOG_REQUEST)
+    public void handleWrite(RequestMsg req, ChannelHandlerContext ctx, IServerRouter r) {
+        LogData logData = getLogData(req.getPayload().getWriteLogRequest().getLogData());
+        log.debug("handleWrite: type: {}, address: {}, streams: {}",
+                logData.getType(), logData.getToken(), logData.getBackpointerMap());
 
         // Its not clear that making all holes high priority is the right thing to do, but since
         // some reads will block until a hole is filled this is required (i.e. bypass quota checks)
         // because the requirement is to allow reads, but only block writes once the quota is exhausted
         if (logData.isHole()) {
-            msg.setPriorityLevel(PriorityLevel.HIGH);
+            req = getRequestMsg(getHighPriorityHeaderMsg(req.getHeader()), req.getPayload());
```

Comment:
I think it would be much cleaner to user the `Builder directly. Also, you should avoid shadowing `req` argument. How about this?

```        
        HeaderMsg.Builder responseHeader = HeaderMsg.newBuilder().mergeFrom(req.getHeader());

        // Its not clear that making all holes high priority is the right thing to do, but since
        // some reads will block until a hole is filled this is required (i.e. bypass quota checks)
        // because the requirement is to allow reads, but only block writes once the quota is exhausted
        if (logData.isHole()) {
            responseHeader.setPriority(CorfuMessage.PriorityLevel.HIGH);
        }

        batchWriter.addTask(BatchWriterOperation.Type.WRITE, req)
                .thenRunAsync(() -> {
                    dataCache.put(logData.getGlobalAddress(), logData);
                    r.sendResponse(getResponseMsg(responseHeader.build(), getWriteLogResponseMsg()), ctx);
                }, executor)
                .exceptionally(ex -> {
                    handleException(ex, ctx, req, r);
                    return null;
                });

Reply:
The issue here is that the message given to `BatchProcessor` when adding the task is expected to have `PriorityLevel.HIGH` when the corresponding `logData` is a hole. In this snippet, we are changing the priority of the response instead. Although this is fine, it still seems like the request needs to be updated in order to use this new header.

Reply:
I have used the `Builder` directly to set `PriorityLevel.HIGH`, but still modify the request message to use this new header.